### DISCUSSION TITLE
Troubleshooting an error on a write

### DISCUSSION THREAD

**Discussion by elitree** [simolus3/drift#2932] [General]
My team has been flummoxed by an error in drift that we're hitting... if anyone's curious and available to look through these details, thanks in advance!

We're seeing an error logged repeatedly in production (via Crashlytics), and having a bunch of trouble recreating the issue locally or even understanding what the stack trace represents. Our production version is using drift v2.12.1, along with Flutter 3.10.7 (we were stuck on that older Flutter release for a while due to a particular iOS bug in 3.13). We see this issue on both iOS and Android.

```
Fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: Null check operator used on a null value
       at DriftCommunication.request(communication.dart:113)
       at _RemoteQueryExecutor.ensureOpen(client_impl.dart:158)
       at LazyDatabase.ensureOpen.<fn>(lazy_database.dart:61)
       at DatabaseConnectionUser._customWrite(connection_user.dart:334)
```

The Flutter error message recorded is one of:
 - Null check operator used on a null value (as seen above)
 - RangeError (length): Invalid value: Not in inclusive range 0..4: 5 (same call stack otherwise)

Here's line 113 in communication.dart in v2.12.1: https://github.com/simolus3/drift/blob/drift-2.12.1/drift/lib/src/remote/communication.dart#L113

Though Crashlytics records this as a "Fatal" error, we don't believe this to be an error that would crash the app--when we see the errors logged, the same user will often have several instances of these errors logged in a row, separated by seconds. And we have had so many instances of this supposedly "fatal" error logged that our service department would have been flooded with reports if they were actual crashes (which we've experienced in the past for other bugs, unfortunately).

Any ideas on what this error might represent or imply? Thanks again for any ideas.

**Comment by simolus3** [comment]
Can you see which dependencies prevent you from using the latest drift version? Drift 2.16 substantially improves the stack trace here. The problem is that the actual error might well be happening on another isolate, which just reports the error back to the one making the request. We capture the stack trace at the time the request was made to provide some context, but that means that you'll see `DriftCommunication.request` at the top of every trace regardless of where the error has actually occurred. In the latest drift version, we're remembering stack traces across isolates and chain them together in the end, leading to a stack trace that should point towards the actual error _and_ the calling context.
So upgrading probably won't fix the issue, but it should give a lot more information to help diagnose it.

> Though Crashlytics records this as a "Fatal" error, we don't believe this to be an error that would crash the app

My guess is that it's probably a write in drift which then throws without being awaited. That error will bubble up and get reported without crashing the app.

**Reply by elitree** [reply]
Sorry for taking so long to write back. 

That's good news that the newer version will help with the stack trace. While we can't update the production version of our app just yet -- our Flutter version in production is stuck back at 3.10, and it looks like we'd need to use 3.19 for a stable compatible Flutter version -- we'll look into upgrading our development version to 3.19 so that we can use Drift 2.16.

Thanks again for the detailed response! I'll write again if I have a further question.
